Source/WebCore/Modules/reporting/ReportingScope.cpp

```diff
@@ -153,17 +153,30 @@ String ReportingScope::endpointURIForToken(const String& reportTo) const
 
 void ReportingScope::generateTestReport(String&& message, String&& group)
 {
-    UNUSED_PARAM(group);
-
+    URL testReportURL;
     String reportURL { ""_s };
-    if (auto* document = dynamicDowncast<Document>(scriptExecutionContext()))
-        reportURL = document->url().strippedForUseAsReferrer();
+
+    auto* document = dynamicDowncast<Document>(scriptExecutionContext());
+    if (document) {
+        testReportURL = document->url();
+        reportURL = testReportURL.strippedForUseAsReferrer();
+    }
+
+    auto testReportBody = TestReportBody::create(WTFMove(message));
 
     // https://w3c.github.io/reporting/#generate-test-report-command, step 7.1.10.
-    auto reportBody = TestReportBody::create(WTFMove(message));
-    notifyReportObservers(Report::create(reportBody->type(), WTFMove(reportURL), WTFMove(reportBody)));
+    if (auto* reportingClient = static_cast<ReportingClient*>(document)) {
```

Comment:
It doesn't make sense that we cast document to its base class here. We should just null check document and use document instead of `reportingClient`. If the issue is that some Document overrides are private, then they should be made public. Why do virtual function calls when we already know the type?

Reply:
I thought that was our typical pattern: Base class is public, children are private. But I have no personal objection to working as you propose here, so I'll make that change.